[Gwyneth Guo Bai Ling] iP

gwynethguo opened the pull request:
### **Kuromi クロミ**

> "I like to write My Melody's mistakes in my notebook. <3" - Kuromi [(source](https://hellokitty.fandom.com/wiki/Kuromi#Personality))

Kuromi クロミ frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/gwynethguo/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😃

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [x] Reminders (coming soon)
---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Kuromi {
    public static void main(String[] args) {
        String home = System.getProperty("user.home");
        java.nio.file.Path path = java.nio.file.Paths.get(home, "Documents", "kuromi.txt");
        new Kuromi(path).run();
    }
}
```

Commit:
Add Gradle support

gwynethguo committed:
Add main features: Greet, Echo, Exit

gwynethguo committed:
Add features: Add, List

gwynethguo committed:
Add feature: Mark As Done

gwynethguo committed:
Add feature: Todo

gwynethguo committed:
Add feature: Deadline

gwynethguo committed:
Add feature: Event

gwynethguo committed:
Automated Text UI Testing: todo, deadline, event, mark, unmark, bye

gwynethguo committed:
Handle errors: todo, mark, unmark, wrong commands

gwynethguo committed:
Add feature: Delete

gwynethguo committed:
Add Save Feature

Duke can receive commands (e.g. todo, deadline, event, bye, mark, etc.)
and save the tasks in a single session.

However, the tasks are not saved after exiting a session (entering
the command "bye").

Adding this feature allows the user to recall the tasks saved before
the current session, by saving the data into the user's hard disk.

gwynethguo committed:
Add Dates Feature

Duke can store the deadline dates as strings for a deadline task.

To make the dates more meaningful, the dates should be stored as dates.

The dates are passed as strings, then it is parsed as date (using
java's LocalDate).

Input from user must be in the format of yyyy-mm-dd. Otherwise, it will
be invalid. Times of the deadline are not yet included in this feature.

gwynethguo committed:
Merge branch 'branch-Level-7'

gwynethguo committed:
Merge branch 'branch-Level-8'

gwynethguo committed:
Fix bugs for Level-8 combined with Level-7

While getting the data from the saved file, a bug will occur if the
type of task is a deadline. This is because the date format in the
saved file is MM dd yyyy instead of yyyy-mm-dd. This will cause the
constructor of the Deadline class to accept MM dd yyyy instead of
yyyy-mm-dd which will cause an error in the constructor.

To fix the error, we should change the date format in the constructor
that is being used to obtain the data from the saved file.

gwynethguo committed:
Reconstruct files to make the code more OOP

The logic of the application is in the Duke.java file, so this does not
follow the OOP Principles: abstraction, encapsulation, polymorphism,
and inheritance.

To follow those principles, we must make separate files and designate
each class to have a specific function (e.g. Parse -> to parse the
user's commands, Ui -> to deal with user interactions, Command -> to
execute the command that the user enters, etc.)

The new classes made are: Ui, Parse, Storage, TaskList, Command
(abstract class), AddCommand, DeleteCommand, ListCommand, MarkCommand,
UnmarkCommand, ExitCommand, CommandType (enum class)

Each class have different functions and the command types are separated
into different classes so as to make the code clear and provide
encapsulation.

Used the enum class for CommandType because there is a finite number
of commands that a user can enter (which makes it suitable according
to the textbook)

gwynethguo committed:
Organize classes into packages to make it easier to find the files for
myself and the collaborator (if a collaborator exists)

gwynethguo committed:
Merge branch 'add-gradle-support'

gwynethguo committed:
Add Gradle Support

gwynethguo committed:
Add JUnit tests

Add JUnit tests to test the behavior of the code.

Add TodoTest and TaskListTest, 2 methods each for unit testing.

gwynethguo committed:
Add Javadoc to enable other people to read the code easily.

gwynethguo committed:
Comply with the Coding Standard

gwynethguo committed:
Add find keyword feature

Users can add, delete, mark, etc. the current tasks.

However, if there are too many tasks, the user will not be able to
find the task that they are searching for.

The find feature allows the user to search for keywords and obtain the
tasks which have the keyword in their description.

This is done by adding find command (class FindCommand) and filter
method in the TaskList class in order to obtain the TaskList that has
been filtered by the keyword.

gwynethguo committed:
Merge branch 'branch-A-JavaDoc'

gwynethguo committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/task/TaskList.java

gwynethguo committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/task/Task.java

gwynethguo committed:
Add changed file: TaskList.java (changes not included in the commit
before)

Huggenguggen commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,67 @@
+        java.nio.file.Path path = java.nio.file.Paths.get(home, "Documents", "duke.txt");
+        new Duke(path).run();
+    }
+}
```
Very clean and neat! :)

Huggenguggen commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,66 @@
+     * Read user's inputs.
+     * @return User input as a string.
+     */
+    public String readCommand() {
+        Scanner scanner = new Scanner(System.in);
+        return scanner.nextLine();
+    }
```
Why do you make a Scanner object every time you call readCommand(). Perhaps you could make 1 Scanner object per Ui object and call Scanner::nextLine() each time you call readCommand() instead?

gwynethguo replied:
I didn't realise that :) Thank you for the input!

gwynethguo committed:
Change Duke to Kuromi.

Change packages, class names to Kuromi (new product name).

gwynethguo committed:
Change some JavaDoc from Duke to Kuromi.

No changes in checkstyle because already imported checkstyle in Week 3.

gwynethguo committed:
Add DialogBox, Launcher, Main, MainWindow for GUI.

There are still some errors occuring for the current commit.

gwynethguo committed:
Edit Commands and DialogBox.

Still have some errors occuring

gwynethguo committed:
Add GUI to Kuromi

To ease users in viewing the chat with Kuromi, a GUI is added.

GUI is added using JavaFX (FXML)

gwynethguo committed:
Update README.md

benjamin-wee commented on src/main/java/kuromi/command/Command.java:
```diff
@@ -0,0 +1,37 @@
+package kuromi.command;
```
I like how u organised your classes into sub packages to make your code more maintainable!

benjamin-wee commented on src/main/java/kuromi/command/MarkCommand.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    /**
+     * Mark the task as done in the task list.
+     * Ask UI to print output.
+     *
+     * @param tasks Task list.
+     * @param ui UI of the application to interact with users.
+     * @param storage Storage to update when there is an update with the task list.
+     */
```
Good job on a highly descriptive javadocs that can help readers understand your code more easily!

benjamin-wee commented on src/main/java/kuromi/Ui.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Deals with interactions with the user (e.g. reading inputs, showing outputs).
+ */
+public class Ui {
```
Good job on following the naming convention of your methods in the Ui class as per the coding standard!

benjamin-wee commented on src/main/java/kuromi/command/ListCommand.java:
```diff
@@ -0,0 +1,24 @@
+     * @param storage Storage to update when there is an update with the task list.
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        String msg = "Got it. I've added this task:\n";
+        msg += tasks.toString() + "\n";
+        msg += "Now you have " + tasks.size() + " tasks in the list.";
```
Good job on following the whitespace guidelines as per the coding standard!

benjamin-wee reviewed (commented):
In general your code is highly readable as it follows the coding standard tightly! Keep up the good work!

gwynethguo committed:
Add welcome message to GUI

gwynethguo committed:
Merge branch 'branch-A-CheckStyle'

gwynethguo committed:
Merge branch 'branch-Level-10'

# Conflicts:
#	src/main/java/kuromi/Kuromi.java
#	src/main/java/kuromi/KuromiException.java

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Add assertions

Kuromi is able to add, delete, get data from storage.

However, there are invariants that should be followed by the
application.

To check if the application follows the invariants, we can add
assertions to ensure that the app follows the rule (e.g. adding a task
increases the task list size by 1, deleting a task decreases the task
list size by 1, and the task list size of a user obtaining tasks from
the storage is the same as the number of tasks in the storage file.

gwynethguo committed:
Align with the Code Quality guidelines

The Parser class contains only one method (parse) that contains the
processes to split the input into description, deadline, etc. for
each command.

This makes the method to be too long (more than 30 lines).

Therefore, I made a method for each command to separate the processes.
The event command might be still too long (around 40 lines), I didn't
find a way to shorten the number of lines for the method.

Furthermore, I changed the execute method for every child classes of
Command class, so that it aligns with SLAP. I also changed the code
in parser method to align with SLAP

gwynethguo committed:
Use Java Streams for find method in TaskList class (for FindCommand)

gwynethguo committed:
Merge pull request #1 from gwynethguo/branch-A-Assertions

Add assertions

gwynethguo committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/kuromi/command/DeleteCommand.java

gwynethguo committed:
Merge branch 'master' into branch-A-Streams

gwynethguo committed:
Merge pull request #2 from gwynethguo/branch-A-CodeQuality

Branch a code quality

gwynethguo committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/kuromi/task/TaskList.java

gwynethguo committed:
Merge pull request #3 from gwynethguo/branch-A-Streams

Branch a streams

gwynethguo committed:
Set up Continuous Integration (CI) (Github Actions)

gwynethguo committed:
Set up Continuous Integration (CI) (GitHub Actions)

gwynethguo committed:
Merge commit '8baea23537a4b8587d3ef7731e758d64b039c725'

gwynethguo committed:
Merge branch 'branch-A-CI'

gwynethguo committed:
Checkstyle issues

gwynethguo committed:
Checkstyle for Assertions Issues

gwynethguo committed:
Change runtest.sh and runtest.bat details

gwynethguo committed:
Change path in runtest.sh and runtest.bat

gwynethguo committed:
Change main class in runtest.sh and runtest.bat

gwynethguo committed:
Merge branch 'master' into branch-A-CI

gwynethguo committed:
Delete I/O Redirection Testing in gradle.yml

gwynethguo committed:
Merge pull request #4 from gwynethguo/branch-A-CI

Branch a ci

gwynethguo committed:
Merge branch 'master' into branch-Level-8

gwynethguo committed:
Add time to deadline dates and datetime to event dates

Catch invalid user input for dates exceptions

gwynethguo committed:
Merge branch 'branch-Level-8'

gwynethguo committed:
Fix checkstyle

gwynethguo committed:
Fix checkstyle for test

gwynethguo committed:
Update test

gwynethguo committed:
More flexibility in searching for items

All more flexibility in search e.g., find items even if the keyword
matches the item only partially.

gwynethguo committed:
Merge pull request #5 from gwynethguo/branch-C-BetterSearch

More flexibility in searching for items

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Catch exceptions for date smaller than current date and start > end date

gwynethguo committed:
Add RemindCommand for the users to be able to ask for Kuromi to remind
them about the first n upcoming tasks.

Todos will be in the least priority because they do not have a deadline.

gwynethguo committed:
Fix checkstyle

gwynethguo committed:
Fix testing issues

gwynethguo committed:
Merge pull request #6 from gwynethguo/branch-B-Reminders

Branch b reminders

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Improve the GUI

Improve the GUI to make it more polished.

Tweak the GUI to match the asymmetric nature of the conversation: As
the conversation is between the user and the app (not between two
humans), it is asymmetric in nature. Accordingly, it makes sense not to
display both sides of the conversion in the same visual format.

Tweak padding, fonts, colors, alignments to make the GUI more pleasing
to look at.

Allow resizing of the Window, and ensure the content resize
appropriately as the Window changes size.

Rounded profile pictures

gwynethguo committed:
Merge pull request #8 from gwynethguo/branch-A-Better-Gui

Improve the GUI

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Give a unique personality to Kuromi

gwynethguo committed:
Merge pull request #9 from gwynethguo/branch-A-Personality

Give a unique personality to Kuromi

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Handle incorrect input for Deadline and Event tasks, and for remind 0
tasks

gwynethguo committed:
Handle NumberFormatException error and add suggestions if typo

gwynethguo committed:
Merge pull request #10 from gwynethguo/branch-A-MoreErrorHandling

Branch a more error handling

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Organise files into packages

gwynethguo committed:
Fix checkstyle

gwynethguo committed:
Add more JUnit tests

gwynethguo committed:
Fix checkstyle

gwynethguo committed:
Merge pull request #11 from gwynethguo/branch-A-MoreTesting

Branch a more testing

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Add help command to list all the commands available

gwynethguo committed:
Add help command and resize the AnchorPane window to square size

gwynethguo committed:
Fix checkstyle

gwynethguo committed:
Add source to MistakesCommand

gwynethguo committed:
Change directory for storage and update .gitignore

gwynethguo committed:
Fix checkstyle

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Add screenshot of the application and fix ExitCommand comments

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Minor changes to the application

Add reminders when launching the application
Change type 'help' to send 'help'
Change help text to follow the format in the user guide

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Update README.md

gwynethguo committed:
Change JavaDoc comments to align with the Java Coding Standard

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Fix checkstyle

gwynethguo committed:
Update README.md

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Fix invalid index error (< 1), fix help command, change message if
keyword not found.

gwynethguo committed:
Update README.md

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip

gwynethguo committed:
Update README.md

gwynethguo committed:
Merge branch 'master' of https://github.com/gwynethguo/ip
